climpred/bootstrap.py

```diff
@@ -234,6 +236,108 @@ def create_pseudo_members(control):
     )
 
 
+def bootstrap_uninit_pm_ensemble_from_control_cftime(init_pm, control):
+    """
+    Create a pseudo-ensemble from control run. Bootstrap random numbers for years to
+    construct an uninitialized ensemble from. This assumes a continous control
+    simulation without gaps.
+
+    Note:
+        Needed for block bootstrapping  a metric in perfect-model framework. Takes
+        random segments of length ``block_length`` from control based on ``dayofyear``
+        (and therefore assumes a constant climate control simulation) and rearranges
+        them into ensemble and member dimensions.
+
+    Args:
+        init_pm (xarray object): ensemble simulation.
+        control (xarray object): control simulation.
+
+    Returns:
+        uninit_pm (xarray object): uninitialized ensemble generated from control run.
+    """
+    check_lead_units_equal_control_time_stride(init_pm, control)
+    # short cut if annual leads
+    if init_pm.lead.attrs['units'] == 'years':
+        return bootstrap_by_stacking(init_pm, control)
+
+    block_length = init_pm.lead.size
+    freq = get_lead_cftime_shift_args(init_pm.lead.attrs['units'], block_length)[1]
+    nmember = init_pm.member.size
+    c_start_year = control.time.min().dt.year.astype('int')
+    # dont resample from years that control wont have timesteps for all leads
+    c_end_year = (
+        shift_cftime_singular(control.time.max(), -block_length, freq).dt.year.astype(
+            'int'
+        )
+        - 1
+    )
+
+    def sel_time(control, start_year_int, suitable_start_dates):
+        """Select time segments from control from ``suitable_start_dates`` based on
+        year ``start_year_int``."""
+        start_time = suitable_start_dates.time.sel(time=str(start_year_int))
+        end_time = shift_cftime_singular(start_time, block_length - 1, freq)
+        new = control.sel(time=slice(*start_time, *end_time))
+        new['time'] = init_pm.lead.values
+        return new
+
+    def create_pseudo_members(control, init):
+        """For every initialization take a different set of start years."""
+        startlist = np.random.randint(c_start_year, c_end_year, nmember)
```

Comment:
Can you make these three all capitals to show that they are constants defined above? `C_START_YEAR`, `C_END_YEAR`, `NMEMBER`.

Reply:
is this necessary? I understand we enforce this for constants that are constant all over climpred. but this is only constant for a given `init`. I think changing that is over the tops and makes the real constants less constant

Reply:
It's constant for a given `init_pm`. These are all defined outside of the function, based on `control` and `init_pm` above. I think it's generally bad practice to reference a global variable within a function, since you could possibly change it here but not have it changed globally. I would either make these required inputs that get passed in, or make them all caps for clarity that they are defined as a constant at the top of the main function, and aren't different for a given iteration. 

Maybe we can ask @ahuang11 for his thoughts on this?

Reply:
Note that Andrew seems to side with you on this case but doesn't really do much subfunctions so didn't have firm thoughts. I think we keep it like this. Only have the variable passed in (`init`) that changes. And know that `c_start_year` and so on can vary but is only varying within the parent function here so it should be fine.